### PR TITLE
perf(django): reduce overheads in instrumentation

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

After CPU profiling, a few areas were discovered in the Django
instrumentation that could account for about 40% CPU overheads. This
change tries to mitigate some of these overheads as much as possible by
avoiding repeating work or optimising some of the tracing steps.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Comment by P403n1x87** [comment]
Current gains:

## Before

~~~
338         8974     176708                       resolver_match = resolver.resolve(request.path_info) 
~~~

## After

~~~
371            .       3494                           resolver_match = request.resolver_match 
~~~

File: ddtrace/contrib/django/patch.py [hunk 1]
```diff
@@ -408,13 +405,15 @@ def traced_get_response(django, pin, func, instance, args, kwargs):
             if response:
                 status = response.status_code
                 span._set_str_tag("django.response.class", func_name(response))
-                if hasattr(response, "template_name"):
+                try:
```
Comment (brettlangdon) [inline ddtrace/contrib/django/patch.py:408]: is `try/except` that much faster than `hasattr()`?

Or is it preferred for another reason?

File: ddtrace/contrib/django/patch.py [hunk 2]
```diff
@@ -400,6 +366,37 @@ def traced_get_response(django, pin, func, instance, args, kwargs):
             # Set HTTP Request tags
             response = func(*args, **kwargs)
 
+            try:
```
Comment (brettlangdon) [inline ddtrace/contrib/django/patch.py:369]: why does moving this improve things?

also, does that mean we can end up with scenarios where we do not resolve the requested url + resource name?

e.g. what happens if `func()` raises an exception (can it?)

File: ddtrace/contrib/django/patch.py [hunk 3]
```diff
@@ -58,14 +58,17 @@
 )
 
 
+config_django = config.django
```
Comment (brettlangdon) [inline ddtrace/contrib/django/patch.py:61]: what impact does this change alone have?

is it really that big of a cost for an attr lookup?

File: ddtrace/contrib/django/utils.py [hunk 4]
```diff
@@ -128,9 +127,9 @@ def get_request_uri(request):
     # combining the url parts. We returns a byte string when all url parts are
     # byte strings.
     # https://github.com/python/cpython/blob/02d126aa09d96d03dcf9c5b51c858ce5ef386601/Lib/urllib/parse.py#L111-L125
-    if PY3 and not all(isinstance(value, binary_type) or value is None for value in urlparts.values()):
```
Comment (brettlangdon) [inline ddtrace/contrib/django/utils.py:131]: cc @majorgreys this was added for a specific reason, no?

Comment (brettlangdon) [reply ddtrace/contrib/django/utils.py:131]: I suppose maybe it doesn't matter anymore if we stop using `urlunparse`?

If so, we can likely remove the comment starting on L123

**Review by brettlangdon — CHANGES_REQUESTED** [review]
it might be better to break up this PR, there are a few disparate things changing, some might be good to 👍🏻 while others we might need to keep discussing

File: ddtrace/contrib/django/patch.py (hunk 1 above)
Comment (P403n1x87) [reply ddtrace/contrib/django/patch.py:408]: As this turns out to be on a hot path, I'd like to squeeze as much as possible, especially if it's a simple change 🙂 .

File: ddtrace/contrib/django/patch.py (hunk 3 above)
Comment (P403n1x87) [reply ddtrace/contrib/django/patch.py:61]: This eats out about 0.1% of the overhead

**Comment by P403n1x87** [comment]
With some caching, the cost of `set_http_meta` has been reduced down to about 1/3. Analysis indicates that other large overheads come from:
- `Span.__exit__` (4.5%)
- `trace_utils.activate_distributed_tracing` (~1.3%)
- `trace_utils.int_service` (~1.2%)

**Comment by P403n1x87** [comment]
This master PR has been split into smaller PRs.

**Pull request closed without merging**